middleware/grpc_cluster.go

```diff
@@ -6,48 +6,122 @@ import (
 
 	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
-	"github.com/gogo/status"
+
+	"github.com/grafana/dskit/clusterutil"
+	"github.com/grafana/dskit/grpcutil"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
-	"google.golang.org/grpc/metadata"
-)
-
-const (
-	MetadataClusterKey = "x-cluster"
+	healthpb "google.golang.org/grpc/health/grpc_health_v1"
 )
 
-// ClusterUnaryClientInterceptor propagates the given cluster info to gRPC metadata.
-func ClusterUnaryClientInterceptor(cluster string) grpc.UnaryClientInterceptor {
+// ClusterUnaryClientInterceptor propagates the given cluster label to gRPC metadata, before calling the next invoker.
+// If an empty cluster label, nil invalidCounter or nil logger are provided, ClusterUnaryClientInterceptor panics.
+// In case of an error related to the cluster label validation, the error is returned, and invalidCounter is incremented.
+func ClusterUnaryClientInterceptor(cluster string, invalidCluster *prometheus.CounterVec, logger log.Logger) grpc.UnaryClientInterceptor {
+	validateClusterClientInterceptorInputParameters(cluster, invalidCluster, logger)
 	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
-		if cluster != "" {
-			ctx = metadata.AppendToOutgoingContext(ctx, MetadataClusterKey, cluster)
+		// We skip the gRPC health check.
+		if method == healthpb.Health_Check_FullMethodName {
+			return invoker(ctx, method, req, reply, cc, opts...)
 		}
 
-		return invoker(ctx, method, req, reply, cc, opts...)
+		ctx = clusterutil.PutClusterIntoOutgoingContext(ctx, cluster)
+		return handleError(invoker(ctx, method, req, reply, cc, opts...), cluster, method, invalidCluster, logger)
+	}
+}
+
+func validateClusterClientInterceptorInputParameters(cluster string, invalidCluster *prometheus.CounterVec, logger log.Logger) {
+	if cluster == "" {
+		panic("no cluster label provided")
+	}
+	if invalidCluster == nil {
+		panic("no invalid cluster counter provided")
+	}
+	if logger == nil {
+		panic("no logger provided")
+	}
+}
+
+func handleError(err error, cluster string, method string, invalidCluster *prometheus.CounterVec, logger log.Logger) error {
```

Comment:
```suggestion
func handleClusterValidationError(err error, cluster string, method string, invalidCluster *prometheus.CounterVec, logger log.Logger) error {
```

We're in the `middleware` package, in such a broad context the method name should be very specific.

Reply:
I have renamed the function. Thank you.